File: management-app-frontend/src/repository/clientRepository.js
import axiosInstance from "../axios/axios.js";

const clientRepository = {
    findAll: async () => {
        return axiosInstance.get("/clients")
    },
    findById: async (id) => {
        return axiosInstance.get(`/clients/${id}`)
    },
    add: async (data) => {
        return axiosInstance.post("/clients/add", data)
    },
    edit: async (id, data) => {
        return axiosInstance.put(`/clients/edit/${id}`, data)
    },
    delete: async (id) => {
        return axiosInstance.delete(`/clients/delete/${id}`)
    },
    findByOrganization: async (organizationId) => {
        return axiosInstance.get(`/clients/organization/${organizationId}`)
    },
    findByFirstName: async(firstName) => {
        return axiosInstance.get("/clients/find-by-first-name", {params: {firstName}})
    },
    findByEmail: async (email) => {
        return axiosInstance.get(`/clients/find-by-email`, {params: {email}})
    }

}
export default clientRepository;

File: ManagementAppBackend/src/main/java/com/managementappbackend/repository/ClientRepository.java
package com.managementappbackend.repository;

import com.managementappbackend.model.domain.Client;
import com.managementappbackend.model.domain.Organization;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Optional;

@Repository
public interface ClientRepository extends JpaRepository<Client, Long> {
    Optional<Client> findByFirstName(String name);
    Optional<Client> findByLastName(String name);
    Optional<Client> findByEmail(String email);
    List<Client> findByOrganization(Organization organization);

    @Query("SELECT c.Id, c.firstName, c.lastName, " +
            "SIZE(c.tasks), " +
            "SUM(CASE WHEN t.status = 'COMPLETED' THEN 1 ELSE 0 END) " +
            "FROM Client c " +
            "LEFT JOIN c.tasks t " +
            "WHERE c.organization = :organization " +
            "GROUP BY c.Id, c.firstName, c.lastName " +
            "ORDER BY SIZE(c.tasks) DESC")
    List<Object[]> findClientTaskDistribution(@Param("organization") Organization organization);

}
